<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>systemPipeShiny • systemPipeShiny</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="systemPipeShiny">
<meta property="og:description" content="systemPipeShiny">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeShiny</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/systemPipeShiny.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/systemPipeR/systemPipeShiny/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="systemPipeShiny_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>systemPipeShiny</h1>
                        <h4 class="author">Author: Le Zhang, Daniela Cassol, Ponmathi Ramasamy, Jianhai Zhang, Gordon Mosher and Thomas Girke</h4>
            
            <h4 class="date">Last update: 29 September, 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/systemPipeR/systemPipeShiny/blob/master/vignettes/systemPipeShiny.Rmd"><code>vignettes/systemPipeShiny.Rmd</code></a></small>
      <div class="hidden name"><code>systemPipeShiny.Rmd</code></div>

    </div>

    
    
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<style>/*Bootstrap colors*/
:root {
  --blue: #007bff;
  --indigo: #6610f2;
  --purple: #6f42c1;
  --pink: #e83e8c;
  --red: #dc3545;
  --orange: #fd7e14;
  --yellow: #ffc107;
  --green: #28a745;
  --teal: #20c997;
  --cyan: #17a2b8;
  --white: #fff;
  --gray: #6c757d;
  --gray-dark: #343a40;
  --primary: #337ab7;
  --secondary: #6c757d;
  --success: #28a745;
  --info: #17a2b8;
  --warning: #ffc107;
  --danger: #dc3545;
  --light: #f8f9fa;
  --dark: #343a40;
  --breakpoint-xs: 0;
  --breakpoint-sm: 576px;
  --breakpoint-md: 768px;
  --breakpoint-lg: 992px;
  --breakpoint-xl: 1200px;
  --font-family-sans-serif: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  --font-family-monospace: SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace;
}

/*
systemPipeShiny CSS Styles
*/

/*buttons, default is too ugly*/
.btn-default {
  color: #fff;
  background-color: var(--primary);
  border-color: #2e6da4;
}

.btn-default:hover{
  background-color: #4c92cf;
}

/*widget user in shinydashboardplus*/
.widget-user {
  overflow-y:auto;
}

/*for clearableTextInput component*/
.text-input-clearable {
  width: 100%;
  border:1px solid #ccc;
  border-radius: 3px;
  padding:1px 6px 1px 1px;
  display:inline-block;
}

.text-input-clearable input {
  width: 93%;
  border:none;
  background:none;
  outline:none;
  padding:0 0;
  margin:0 0;
  font:inherit;
}

.text-input-clearable span {
  float: right;
  height: 14px;
  font-size: 14px;
  cursor: pointer;
  color: #ccc;
  visibility:hidden;
}
.text-input-focused {
  border-color: #66afe9;
  outline: 0;
  -webkit-box-shadow: inset 0 1px 1px rgba(0,0,0,.075), 0 0 8px rgba(102, 175, 233, 1);
  -moz-box-shadow: inset 0 1px 1px rgba(0,0,0,.075), 0 0 8px rgba(102, 175, 233, 1);
  box-shadow: inset 0 1px 1px rgba(0,0,0,.075), 0 0 8px rgba(102, 175, 233, 1);
}

/*for gallery */
img-gallery{
  max-width:100%;
  height:auto;
  object-fit: cover;
  vertical-align: middle;
  text-align: center;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
}

img-gallery:hover {
  opacity: 0.7;
  transition: opacity 0.5s;
}


/*for sps hyperlink button  and tables*/
.href-button {
  background-color: var(--primary);
  border: none;
  border-radius: 10px 10px;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
  color: white;
  padding: 10px 25px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  font-size: 15px;
  margin: 4px 2px;
  cursor: pointer;
         }

.href-button:hover {
  opacity: 0.9;
  text-decoration: none;
  color: white;
  transition: opacity 0.5s;
}

.href-button:active {
  transform: translateY(2px);
}

.table-href td:first-child{
  width: 1px;
  vertical-align: middle;
  white-space: nowrap;
}

/* about tab background */
.widget-user-header {
    background: linear-gradient(-45deg, #428bca, #ccccff, #b3b3b3);
    background-size: 400% 400%;
    animation: gradient 10s ease infinite;
}

@keyframes gradient {
    0% {
        background-position: 0% 50%;
    }
    50% {
        background-position: 100% 50%;
    }
    100% {
        background-position: 0% 50%;
    }
}

/* sps description container*/
.desc-body.collapse {
  display: block;
  height: 150px;
  overflow: hidden;
}
.desc-body.in {
  height: auto;
}
.desc a.collapsed:after  {
    content: '+ Show More Description';
}
.desc > a:not(.collapsed):after {
    content: '- Collapse Description';
}
.desc a:not(.desc-body) {
  font-size: 1.1em;
}
.desc a:not(.desc-body):link {
  text-decoration: none;
}

/* sps control-panel*/
.control-panel {
    opacity: 0.7;
    transition: opacity .25s ease-in-out;
   -moz-transition: opacity .25s ease-in-out;
   -webkit-transition: opacity .25s ease-in-out;
}

.control-panel:hover { opacity: 1;}


/* sps hexagon icon and hex panel*/
.hex-container {
  display: inline-block;
  text-align: center;
  width: 100%;
}

.hex {
  opacity: 0.8;
}

.hex:hover {
  opacity: 1;
}

.powerby-link {
  font: italic bold 25px serif;
  fill: var(--primary);
  opacity: 0.5;
}

.powerby-link:hover {
  opacity: 1;
}

.hex-box {
  width: 3vw;
  min-width: 50px;
  display: inline-block;
  padding-top: 0;
}

.hex-container span {
 display: inline-block;
 text-align: center;
 font-size: 0.6vw;
}

.hex-panel {
  position: auto;
}
.hex-panel > .hex-item {
  display: inline-block;
  text-align: left;
  padding-left: 0;
  padding-right: 0;
  margin-left: 0;
  margin-right:0;
}

.hex-panel h5 {
  display: inline-block;
  margin-left: 20px;
  font-weight: 900;
  font-size: 1em;
  transform: translateY(-2.2vw);
}

/* sps canvas*/
.snap-drag {
  opacity: 0;
}

.snap-drag:hover {
  opacity: 1;
}

.snap-drag button{
  position: absolute;
  outline: none;
  background-color: Transparent;
  top: 2px;
  right: 4px;
}

.snap-drag button:focus {
  outline: 0 !important;
  box-shadow: none !important;
}

.snap-drag h4{
  margin-bottom: 0;
  padding-bottom: 10px;
}

a.bttn-default {
    background: #3c8dbc !important;
}

.sps-warning h4 {
    color: var(--danger);
}
.sps-warning h4::before{
    content: "\2022 ";
}
</style>
<p><img src="../inst/app/www/img/sps.png" width="70%" style="display: block; margin: auto;"></p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><strong><span style="color:#5DA7D6;">s</span>ystem<span style="color:#5DA7D6;">P</span>ipe<span style="color:#5DA7D6;">S</span>hiny</strong>(SPS) is a <a href="https://shiny.rstudio.com/">Shiny</a>-based R/Bioconductor package that has a graphical user interface (GUI). SPS is able to:</p>
<ol style="list-style-type: decimal">
<li>Help for researchers on data analysis workflow management and downstream data visualizations in quick and user friendly manner.</li>
<li>Provides an framework to be fast extended with more components and an web app that is open-box-ready-to-deploy to any Shiny server.</li>
</ol>
<p>You can try our demo at <a href="https://tgirke.shinyapps.io/systemPipeShiny/">here</a>. There are 5 different interactive tutorials embedded on the demo that you can follow(right upper corner). Each tutorial covers a major topic of SPS and shows how you can interactive with app.</p>
<p>This vignette illustrate the main concepts of SPS by using an RNAseq workflow as the example.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>To install SPS:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"systemPipeShiny"</span><span class="op">)</span></pre></div>
<p>If you are on Linux, you also need(use Ubuntu as example):</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">sudo</span> apt-get install libv8-dev</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="fu">sudo</span> apt-get install libssl-dev</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="fu">sudo</span> apt-get install libcurl4-openssl-dev</span></code></pre></div>
<p><em>libcurl4-openssl-dev</em> may not be required for Ubuntu &gt; 18.04.</p>
<p>Depending on the Linux distributions, the command can be slightly different.</p>
</div>
<div id="main-concept" class="section level1">
<h1 class="hasAnchor">
<a href="#main-concept" class="anchor"></a>Main concept</h1>
<p>The SPS is designed into two major categories, workflow management (WF) and data visualization (VS) (<em>Fig 1</em>). WF part helps users to prepare data analysis workflow files. VS part helps users to make different types of plots, take interactive snapshot of plots and combine/compare different plots on the “Canvas” workbench.</p>
<center id="sps_structure">
<img src="img/sps_structure.png" alt="SPS_structure"><strong>Figure 1. SPS Structure</strong>
</center>
<p>Inside the two major categories, there are small components. They are <strong>SPS tabs</strong>. A SPS tab is similar to any other web dashboard tabs you see on the internet. Tabs in SPS are built on top of <a href="https://shiny.rstudio.com/articles/modules.html">Shiny modules</a> as individual files. The modular structure makes it very easy to extend with more SPS tabs. Functions and instructions are provided in the <a href="#extend-sps">developer section</a>.</p>
</div>
<div id="step-by-step-instruction-of-with-an-example-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#step-by-step-instruction-of-with-an-example-workflow" class="anchor"></a>Step-by-step instruction of with an example workflow</h1>
<p>Let us use an <strong>RNAseq</strong> workflow as the example to demonstrate how to use SPS.</p>
<div id="create-a-sps-project" class="section level2">
<h2 class="hasAnchor">
<a href="#create-a-sps-project" class="anchor"></a>Create a SPS project</h2>
<p>After installation and before launching SPS, the first thing you need to do is to create a SPS project:</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">sps_tmp_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="../reference/spsInit.html">spsInit</a></span><span class="op">(</span>app_path <span class="op">=</span> <span class="va">sps_tmp_dir</span>, change_wd <span class="op">=</span> <span class="cn">FALSE</span>, project_name <span class="op">=</span> <span class="st">"SPSProject"</span><span class="op">)</span></pre></div>
<pre><code>## [SPS-INFO] 2020-09-29 14:19:18 Start to create a new SPS project</code></pre>
<pre><code>## [SPS-INFO] 2020-09-29 14:19:18 Create project under /tmp/RtmpqPgcj5/SPSProject</code></pre>
<pre><code>## [SPS-INFO] 2020-09-29 14:19:18 Now copy files</code></pre>
<pre><code>## [SPS-INFO] 2020-09-29 14:19:18 Create SPS database</code></pre>
<pre><code>## [SPS-INFO] 2020-09-29 14:19:18 Created SPS database method container</code></pre>
<pre><code>## [SPS-INFO] 2020-09-29 14:19:19 Creating SPS db...</code></pre>
<pre><code>## [SPS-DANGER] 2020-09-29 14:19:19 Db created at '/tmp/RtmpqPgcj5/SPSProject/config/sps.db'. DO NOT share this file with others
## [SPS-INFO] 2020-09-29 14:19:19 Key md5 27f687f41b92ce1b06fae07112675ad1
## [SPS-INFO] 2020-09-29 14:19:19 SPS project setup done!</code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="va">sps_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">sps_tmp_dir</span>, <span class="st">"SPSProject"</span><span class="op">)</span></pre></div>
<p>Here for building the vignette we are not switching to the app directory and are using a temp directory, in a real case, you <strong>should not</strong> store you SPS project in a <strong>temp directory</strong>. Just use the following instead:</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="fu"><a href="../reference/spsInit.html">spsInit</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<p>By default, you will have a folder created under your current working directory, the name is <code>SPS_</code> + <code>today's date</code> in <em>ISO 8601</em> format, e.g.  SPS_20200929. It looks like this:</p>
<pre>
SPS_xx/  
├── server.R               <span class="text-success">|</span> 
├── global.R               <span class="text-success">| Most important server, UI and global files, unless special needs, `global.R` is the only file you need to edit manually</span>   
├── ui.R                   <span class="text-success">|</span>  
├── deploy.R               <span class="text-info">| Deploy helper file</span>  
├── config                 <span class="text-success">| Important app config files. Do not edit them if you don't know</span>  
│   ├── sps.db             <span class="text-gray">| SPS database</span> 
│   ├── sps_options.yaml   <span class="text-gray">| SPS default option list</span> 
│   └── tabs.csv           <span class="text-gray">| SPS tab information</span> 
├── data                   <span class="text-primary">| App example data files</span> 
│   ├── xx.csv             
├── R                      <span class="text-primary">| All SPS additional tab files and helper R function files</span> 
│   ├── tab_xx.R            
├── README.md              
├── results                <span class="text-gray">| Not in use for this current version, you can store some data been generated from the app</span> 
│   └── README.md          
└── www                    <span class="text-primary">| Internet resources</span>  
    ├── about              <span class="text-gray">| About tab information</span> 
    │   └── xx.md          
    ├── css                <span class="text-info">| CSS files</span>  
    │   └── sps.css         
    ├── img                <span class="text-info">| App image resources</span>    
    │   └── xx.png         
    ├── js                 <span class="text-info">| Javascripts</span>
    │   └── xx.js           
    ├── loading_themes     <span class="text-info">| Loading screen files</span> 
    │   └── xx.html         
    └── plot_list          <span class="text-info">| Image files for plot gallery</span>  
        └── plot_xx.jpg      
</pre>
</div>
<div id="launch-sps" class="section level2">
<h2 class="hasAnchor">
<a href="#launch-sps" class="anchor"></a>Launch SPS</h2>
<p>By default, your working directory should be set inside that project folder automatically. If you are using Rstudio, <em>global.R</em> file will be opened for you. This is the only file that you need to make custom change if there is any.</p>
<p>In your <em>global.R</em>, scroll down to the bottom, you should see:</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="va">sps_app</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sps.html">sps</a></span><span class="op">(</span>vstabs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"data_example"</span>, <span class="st">"plot_example1"</span>, <span class="st">"plot_example2"</span><span class="op">)</span>, 
    plugin <span class="op">=</span> <span class="st">""</span>, server_expr <span class="op">=</span> <span class="op">{</span>
        <span class="fu"><a href="../reference/msg.html">msg</a></span><span class="op">(</span><span class="st">"Custom expression runs -- Hello World"</span>, <span class="st">"GREETING"</span>, 
            <span class="st">"green"</span><span class="op">)</span>
    <span class="op">}</span><span class="op">)</span></pre></div>
<p>This is the SPS main function. You can load/unload tabs by providing tab IDs in <code>vstabs</code> argument, like <code>c("tab1", "tab2)</code>. Open <em>config/tabs.csv</em> or use <code><a href="../reference/spsTabInfo.html">spsTabInfo()</a></code> to see what tabs IDs can be load and other tab information. Currently you can only load/unload visualization tabs, the key word <code>vs</code> under column “<em>type</em>”; essential framework tabs(<em>core</em>) and workflow tabs (<em>wf</em>) are loaded automatically and these tabs cannot be modified. The <a href="#extend-sps">developer section</a> will teach you how to create your own tabs.</p>
<p>For this vignette example, we will use as it is, so now you can run the SPS by typing <code><a href="https://shiny.rstudio.com/reference/shiny/latest/runApp.html">shiny::runApp()</a></code> in console or click on the <img src="img/app_start.png" style="display: inline; padding-left:0; width:10px; height:15px;"><code>Run App</code> button on top right corner in Rstudio. The app should be launched. Click on the “Continue to app” button on the welcome screen and you should see SPS main dashboard like this one (Fig.2):</p>
<center id="sps_ui">
<img src="img/sps_ui_main.jpg" alt="sps_ui"><strong>Figure 2. SPS UI</strong>
</center>
<p>All SPS tabs are well classified into WF and VS part. We will start from WF management.</p>
</div>
<div id="wf-management" class="section level2">
<h2 class="hasAnchor">
<a href="#wf-management" class="anchor"></a>WF management</h2>
<p>WF management in SPS helps you to prepare a data analysis workflow files in <a href="https://systempipe.org/docs/systemPipeR/">systemPipeR</a> (SPR) format. SPR is another R package that runs a WF in “command line + R code” manner. To run a SPR WF, you need to prepare three very important files: a metadata(targets) file that contains sample information, a workflow step file(Rmarkdown format) that defines the WF steps and a WF config file that helps the program to set up your directory. In SPS, one can easily prepare all of these three files in SPS tabs under “Workflow Management” section. You can find them at the left side navigation bar (Fig2 a-b).</p>
<p>Under current version of SPS, we only help you to prepare and check format of these files. One should save these files and run a WF in SPR. Directly running a SPR WF in SPS will be supported in an upcomming release.</p>
<div id="prepare-a-target-file" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-a-target-file" class="anchor"></a>Prepare a target file</h3>
<p>Read <a href="https://systempipe.org/docs/systemPipeR/#structure-of-targets-file">“Structure of targets file”</a> section in SPR vignette to better undertand the structure of this file.</p>
<p>Now let us follow the order below to prepare a target file for a RNAseq project (Fig.3 A):</p>
<ol style="list-style-type: decimal">
<li>Choose “Targets” from navigation bar;</li>
<li>Select either “Example PE” or “Example SE”" for using a pair-end/single-end target file, then choose “confirm” on the pop-up. You will see the exmaple is loaded like Figure 3A.</li>
<li>You can edit, right click to add/remove rows/columns (The first row is treated as column names);</li>
<li>SPR target file requres a header and you can edit it here. Headers should start with “#”;</li>
<li>The left section can help you to check sample statistics and file existence. Choose any column you want from the dropdown to check and watch the statistics bar change in this section;</li>
<li>Click on “Add to task” can help you to check if your target file has any formatting problem. You should see a green success pop-up if everything is right;</li>
<li>Now your target file is ready and you can click “save” to download it and later use in a SPR project.</li>
<li>Click on the “-” button on the “Workflow Progress”, you should see the “Targets file” becomes green.</li>
</ol>
<center>
<img src="img/sps_ui_targets.png" alt="wf_targets"><strong>Figure 3. A. Workflow Management - Targets File</strong>
</center>
</div>
<div id="prepare-a-workflow-file" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-a-workflow-file" class="anchor"></a>Prepare a workflow file</h3>
<p>In SPR, workflows are defined in Rmarkdown files, you can read details and obtain them <a href="https://systempipe.org/pages/pipelines_area/">here</a>.</p>
<p>Now let us follow the order below to see how SPS helps you to prepare a WF file for a RNAseq project (Fig.3 B):</p>
<ol style="list-style-type: decimal">
<li>Choose “Workflow File” from navigation bar;</li>
<li>Choose “Example Rmd” to load an RNAseq WF;</li>
<li>You can see the WF steps are displayed like a tree structure;</li>
<li>Check all steps in the WF that you want to include. You can skip (leave blank) some steps but it may cause the WF to fail. Read more SPR instructions before do so.</li>
<li>Click on the “Plot steps” will show a flow chart of what the step execution orders will be when you run the WF in SPR.</li>
<li>Click “Report preview” generates a preview of what the final report will look like for current RNAseq WF, but in the preview, no code is evaluated.</li>
<li>If you are satisfied with your WF file, click “Add to task” to update the workflow preparation progress just like it is for “targets file”.</li>
<li>Now choose “Save New Rmd” to save this new WF file to later be used a SPR RNAseq project running.</li>
</ol>
<center>
<img src="img/sps_ui_wf.png" alt="wf_wf"><strong>Figure 3. B. Workflow Management - Workflow File</strong>
</center>
</div>
<div id="prepare-a-configuration-file" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-a-configuration-file" class="anchor"></a>Prepare a configuration file</h3>
<p>In SPR, the config file <em>SPRconfig.yaml</em> is required to define information like <em>project path</em>, <em>targets file path</em> and more. Normally, this file is generated automatically when one starts a SPR project, but if any modification is made, one needs to update this config file. For example, like we are now using SPS to prepare targets file and WF file and when we download these new files, file names and locations may be changed. In this case, we can update the config file in SPS (Fig.3 C):</p>
<ol style="list-style-type: decimal">
<li>Choose “Targets” from navigation bar;</li>
<li>Choose “Example” to load an RNAseq config file;</li>
<li>Edit the file content and follow <em>yaml</em> format. You may need to update the target file and WF file name and location in the “<em>targets</em>” and “<em>Rmd</em>” lines here.</li>
<li>Once your editing is done, click “Add to task” to update the progress;</li>
<li>Now your new config file is ready and you can click “Save” to download it;</li>
<li>The progress tracker should all become green, which means you are ready to run the RNAseq workflow in SPR.</li>
</ol>
<center>
<img src="img/sps_ui_config.png" alt="wf_config"><strong>Figure 3. C. Workflow Management - Config File</strong>
</center>
</div>
<div id="finish-wf-management" class="section level3">
<h3 class="hasAnchor">
<a href="#finish-wf-management" class="anchor"></a>Finish WF management</h3>
<p>If you have followed this instruction so far, the three most important WF files have been created. You can copy them over to your SPR RNAseq project root and start to generate results.</p>
</div>
</div>
<div id="visualize-workflow-results" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-workflow-results" class="anchor"></a>Visualize workflow results</h2>
<p>In the previous step that we have prepared files for running a SPR RNAseq workflow. Now assume that the workflow has been successfully run and we have obtained some results. In this part, we will use an example <strong>count table</strong> (RPKM normalized) to demonstrate how to use the VS components.</p>
<p>VS tabs in SPS are sub-defined into two small groups, <em>data tabs</em> and <em>plot tabs</em>. A data tab is where you prepare, filter and preprocess data. A plot tab is where the actual plots are generated. A plot tab requires at least one dataset from a data tab. Therefore, we need to first prepare data in a data tab, depending on the preprocess method we choose, we will be given different plot options (going to different plot tabs).</p>
<p>In this part, we will first load the raw count table and preprocess it in “Data Tab Example”, generate some plots in “Plot Tab Example”, and finally compare plots in the “Canvas” tab.</p>
<div id="load-raw-count-data" class="section level3">
<h3 class="hasAnchor">
<a href="#load-raw-count-data" class="anchor"></a>Load raw count data</h3>
<p>Follow instructions in Figure 4A:</p>
<ol style="list-style-type: decimal">
<li>Select “Data Tab Example” tab from the left;</li>
<li>Click “Start with this tab”, it will be hidden after package requirement check has passed in the background. So it is not displayed in the figure.</li>
<li>Let us load an example RPKM read count table. If you upload your own table, the first column must be “Genes” and the rest of columns are treated as samples and they must be numeric values only.</li>
<li>You can click on box below each column name to filter data.</li>
<li>You can export your filtered data into clipboard, csv files, Excel …</li>
<li>After the data filtration, we can start to preprocess data. Here are two methods you can choose from the left dropdown menu: “Do Nothing” does not change the filtered data, and “PCA” calculates PCs based on samples. Let us just try “Do Nothing”;</li>
<li>When the data is preprocessed, plot options will be displayed in the below. You need to scroll down a little to see. “Do Nothing” will give you the option for a bar plot. Click on the plot image will redirect you to the corresponding tab.</li>
</ol>
<center>
<img src="img/sps_ui_data_tab.png" alt="data_tab"><strong>Figure 4 A. VS - data tab</strong>
</center>
</div>
<div id="create-a-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#create-a-plot" class="anchor"></a>Create a plot</h3>
<p>Follow instructions in Figure 4B:</p>
<ol style="list-style-type: decimal">
<li>Skip this step if you jumped directly from the “Data Tab Example”, otherwise select “Plot Bar” to start;</li>
<li>Click “Start/Reload”, this button will validate if you have the required data set. If yes, you will see plot options and other UI.</li>
<li>Click on “Render/Snapshot plot” to create the first plot. In SPS, the first time clicking on “Render” button creates a preview of the plot. From the second time and on, clicking on this button will take a <em>“snapshot”</em> of the plot. In SPS a snapshot is plot frozen with particular set of input data and plot options. It is isolated from the original plot tab and you cannot modify it anymore. It is similar to a screenshot but you can still interact within the plot. You will have a better understanding when we reach the <em>Canvas</em> section.</li>
<li>Now you can interact with plot. Try to hover your mouse and move around the plot. There are many interactive components on the plot. Read more of what you can interact with a <a href="plotly.com">Plotly plot</a>.</li>
<li>There is a tiny triangular icon on the right bottom corner of the plot where you can drag it to resize the plot to at most your window width and unlimited height;</li>
<li>A panel of plot selection tools including a screenshot button is on the top right;</li>
<li>Let us take a snapshot by clicking “Render”. You should see a blue message pop up from the right bottom of your screen.</li>
<li>You can change the plot title by changing the text. This example plot tab only provides very simple plot controls. Change the text and watch how the plot title change.</li>
<li>Change a title and click “Render” again to take a second snapshot.</li>
</ol>
<center>
<img src="img/sps_ui_plot_tab.png" alt="plot_tab"><strong>Figure 4 B. VS - plot tab</strong>
</center>
</div>
</div>
<div id="snap-shots-and-canvas" class="section level2">
<h2 class="hasAnchor">
<a href="#snap-shots-and-canvas" class="anchor"></a>Snap shots and Canvas</h2>
<div id="canvas" class="section level3">
<h3 class="hasAnchor">
<a href="#canvas" class="anchor"></a>Canvas</h3>
<p>SPS Canvas is a place to display snapshots. Snap shots can come from different plot tabs, or the same plot tab with different plot options. Here you can combine plots together to a big scientific figure and compare plots with different plotting options side by side.</p>
<p>If you have followed the instructions so far, you should have at least two snapshots and we can start to use Canvas.</p>
<ol style="list-style-type: decimal">
<li>Select “Canvas” tab;</li>
<li>You will see your snapshots in “Current snapshots” box. If you do not want to see any of them, drag it to the right side so it will not be rendered on Canvas.</li>
<li>Click “Refresh Canvas” to render the Canvas for a quick glimpse;</li>
<li>You can change some Canvas settings on top;</li>
<li>In the “bar plot” section, we took two snapshots, one with original plot title and one with custom title. If you made the change, you can see it is reflected on Canvas. Move your mouse on the plots, you can see they are still interactive. Then, when you move the mouse near the top of a plot, a title section will show up and you can drag it to reposition your plot.</li>
<li>Similar to a plot tab, you can resize any plot on Canvas by dragging the bottom right corner of a plot.</li>
</ol>
<center>
<img src="img/sps_ui_canvas.png" alt="plot_tab"><strong>Figure 5 A. Canvas</strong>
</center>
<p>Now go back to <a href="#load-raw-count-data">Load raw count data</a> section and select “PCA” option in step 6 to reprocess and then go to PCA plot tab and repeat instructions in <a href="#create-a-plot">Create a plot</a> to create and snapshot a few PCA plots. The PCA plot tab has an additional plot point size control UI. You can try to take snapshots with different point size. Finally come back to canvas and render all snapshots. You should then be able to see how point size on plots are different in different snapshots. Now you can have a better understanding of the SPS snapshot-Cnavas feature.</p>
</div>
<div id="snapshot-management" class="section level3">
<h3 class="hasAnchor">
<a href="#snapshot-management" class="anchor"></a>Snapshot management</h3>
<p>Taking too many snapshots will slow down the app. A good solution is to manage them through the snapshot manage menu. You can access it by clicking the button “Snapshots” with a camera icon on the very top of SPS dashboard banner. It will open up a big menu screen (Fig. 5B).</p>
<ol style="list-style-type: decimal">
<li>The left side displays all current snapshots. The label means the plot tab ID of which plot tab it comes from and the number after the dash means it is the Nth snapshot coming from that particular plot tab;</li>
<li>If you want to delete any snapshot, drag it to the right;</li>
<li>Click on “DESTROY” will permanently delete selected snapshots and cannot be recovered.</li>
<li>Download your current snapshots to a compressed file.</li>
<li>You can upload saved snapshot file and continue to work on it later.</li>
</ol>
<p>Due to internet security reasons, snapshot files are encrypted. Each SPS project has an unique key pair. Therefore, it is not possible to run a snapshot file in a different SPS project.</p>
<center>
<img src="img/sps_ui_snap.png" alt="plot_tab"><strong>Figure 5 B. snapshot management</strong>
</center>
</div>
</div>
</div>
<div id="advanced-features" class="section level1">
<h1 class="hasAnchor">
<a href="#advanced-features" class="anchor"></a>Advanced features</h1>
<div id="sps-plugins" class="section level2">
<h2 class="hasAnchor">
<a href="#sps-plugins" class="anchor"></a>SPS plugins</h2>
<p>We have created many more plots options and they are distributed as “SPS plugins”. A SPS plugin is a collection of SPS tabs that become small R packages. You need to install the plugin like you do with normal R packages by <code>install.packages</code>, <code><a href="https://rdrr.io/pkg/BiocManager/man/install.html">BiocManager::install</a></code> or <code><a href="https://remotes.r-lib.org/reference/install_github.html">remotes::install_github()</a></code>. The plugins(packages) come with SPS tabs, data, and other helper functions. Using <code><a href="../reference/spsAddPlugin.html">spsAddPlugin()</a></code> function, you are able to load them into SPS framework.</p>
<p>Using <code><a href="../reference/spsAddPlugin.html">spsAddPlugin()</a></code> without any argument will list all available plugins and instructions to install them.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="fu"><a href="../reference/spsAddPlugin.html">spsAddPlugin</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<pre><code>## [SPS-INFO] 2020-09-29 14:19:19 Current SPS has following plugins</code></pre>
<pre><code>## spsBio               Biological plots
## BiocManager::install("systemPipeR/spsBio")</code></pre>
<p>To load a specific plugin:</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="fu"><a href="../reference/spsAddPlugin.html">spsAddPlugin</a></span><span class="op">(</span>plugin <span class="op">=</span> <span class="st">"PLUGIN_NAME"</span>, app_path <span class="op">=</span> <span class="st">"PATH_OF_SPS"</span><span class="op">)</span></pre></div>
<p>At this stage, there is only one plugin, the <em>spsBio</em>: plots for biological data visualization. It is <strong>under testing and please expect unstableness</strong>. Under your project directory:</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"systemPipeR/spsBio"</span><span class="op">)</span>
<span class="fu"><a href="../reference/spsAddPlugin.html">spsAddPlugin</a></span><span class="op">(</span>plugin <span class="op">=</span> <span class="st">"spsBio"</span><span class="op">)</span></pre></div>
<p>Inside your SPS project root, in the <em>global.R</em> file, specify that you want to load the plugin to app and start the app like normal:</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="va">sps_app</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sps.html">sps</a></span><span class="op">(</span>vstabs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"data_example"</span>, <span class="st">"plot_example1"</span>, <span class="st">"plot_example2"</span><span class="op">)</span>, 
    plugin <span class="op">=</span> <span class="st">"spsBio"</span>, server_expr <span class="op">=</span> <span class="op">{</span>
        <span class="fu"><a href="../reference/msg.html">msg</a></span><span class="op">(</span><span class="st">"Custom expression runs -- Hello World"</span>, <span class="st">"GREETING"</span>, 
            <span class="st">"green"</span><span class="op">)</span>
    <span class="op">}</span><span class="op">)</span></pre></div>
</div>
<div id="configuration" class="section level2">
<h2 class="hasAnchor">
<a href="#configuration" class="anchor"></a>Configuration</h2>
<div id="app-options" class="section level3">
<h3 class="hasAnchor">
<a href="#app-options" class="anchor"></a>App options</h3>
<p>App options in SPS are controlled by “SPS options”. These options can change app appearance, debugging level, server server behaviors, <em>etc</em>. The valid options can be found and change on the <code>global.R</code> file. They are similar to <em>Shiny</em> options, but unlike <em>shiny</em> options are single values, SPS options are passed using the <code>Option(sps = list(...))</code> function in <code>global.R</code> as a group. To view all options and their default, valid values(’*’ means any value is okay):</p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="fu"><a href="../reference/viewSpsDefaults.html">viewSpsDefaults</a></span><span class="op">(</span>app_path <span class="op">=</span> <span class="va">sps_dir</span><span class="op">)</span></pre></div>
<pre><code>## mode: list(default = "local", other = "server")
##  loading_screen: list(default = TRUE, other = FALSE)
##  loading_theme: list(default = "vhelix", other = "*")
##  use_crayon: list(default = TRUE, other = FALSE)
##  verbose: list(default = FALSE, other = TRUE)
##  admin_page: list(default = FALSE, other = TRUE)
##  admin_url: list(default = "admin", other = "*")
##  warning_toast: list(default = TRUE, other = FALSE)</code></pre>
<p>To view a single option value, use <code><a href="../reference/spsOption.html">spsOption(opt = "OPTION_NAME")</a></code>; to overwrite a single option, use <code><a href="../reference/spsOption.html">spsOption(opt = "OPTION_NAME", value = "NEW_VALUE")</a></code>.</p>
<p style="color:var(--info)">
Note: Do not worry if you set some invalid values, on app start, <code><a href="../reference/sps.html">sps()</a></code> will check all SPS options, ignore unknown values and set invalid values back to default. You will see warning messages on console to tell you specifically what is wrong with your options.
</p>
<table class="table">
<thead><tr class="header">
<th>Option</th>
<th>Description</th>
<th>Default</th>
<th>Other</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>mode</td>
<td>running mode</td>
<td>“local”</td>
<td>“server”</td>
</tr>
<tr class="even">
<td>warning_toast</td>
<td>show security warnings?</td>
<td>TRUE</td>
<td>FALSE</td>
</tr>
<tr class="odd">
<td>loading_screen</td>
<td>show loading screen?</td>
<td>TRUE</td>
<td>FALSE</td>
</tr>
<tr class="even">
<td>loading_theme</td>
<td>loading screen theme</td>
<td>“vhelix”</td>
<td>any string</td>
</tr>
<tr class="odd">
<td>use_crayon</td>
<td>colorful console message?</td>
<td>TRUE</td>
<td>FALSE</td>
</tr>
<tr class="even">
<td>verbose</td>
<td>more details for SPS functions?</td>
<td>TRUE</td>
<td>FALSE</td>
</tr>
<tr class="odd">
<td>admin_url</td>
<td>admin_page query url</td>
<td>“admin”</td>
<td>any string</td>
</tr>
</tbody>
</table>
<p>some details:</p>
<ul>
<li>mode: this option will change how the upload files are selected.
<ul>
<li>local: You can choose file from server side, and it is a direct link, no file will be copied to server. Most often used when you are on you own computer. <span style="color:var(--info)">It is still choosing files from the server side, because you are using your own computer as a Shiny server.</span>
</li>
<li>server: Choose files from the client side, and upload as a temp file to the server.</li>
<li>If you are on you own computer, there is no difference, but <code>local</code> mode can avoid copy your files to temp. <span style="color:var(--info)">If you are deploying the app to cloud, please <strong>use server mode</strong></span>
</li>
</ul>
</li>
<li>warning_toast: A toast pop-up message when you are under some dangerous options when app starts.</li>
<li>loading_theme: <code>loading_screen</code> must be <code>TRUE</code>. Only one theme for now, we will add more in future.</li>
<li>verbose: Give you more information on debugging.</li>
<li>admin_url: the admin page url , will be added in an upcoming version, a preview can be seen by adding “/?admin” to the end of app url, e.g. “<a href="https://tgirke.shinyapps.io/systemPipeShiny/?admin">https://tgirke.shinyapps.io/systemPipeShiny/?admin</a>”.</li>
</ul>
</div>
<div id="tab-configurations" class="section level3">
<h3 class="hasAnchor">
<a href="#tab-configurations" class="anchor"></a>Tab configurations</h3>
<p>Tabs are registered(defined) by the <em>tabs.csv</em> file under SPS project <em>config</em> folder. It contains information for all the tabs. Do not modify it if you don’t plan to build new tabs. See <a href="#extend-spsfor-developers">developer section</a> for how to operate this file.</p>
<p>To control what tabs to appear(load/unload) on SPS, pass the tab IDs of tabs you want to see to <code>vstabs =</code> argument on <code><a href="../reference/sps.html">sps()</a></code></p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="va">sps_app</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sps.html">sps</a></span><span class="op">(</span>vstabs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ID1"</span>, <span class="st">"ID2"</span>, <span class="va">...</span><span class="op">)</span><span class="op">)</span></pre></div>
<p><em>tabs.csv</em> has all the tab IDs. Under current version of SPS, you can only load/unload tabs in the “vs” type. Tabs under “core” and “wf” types are unable to change.</p>
</div>
</div>
<div id="a-variety-of-additional-shiny-widgets" class="section level2">
<h2 class="hasAnchor">
<a href="#a-variety-of-additional-shiny-widgets" class="anchor"></a>A variety of additional shiny widgets</h2>
<p>There are many useful HTML components that can be use outside out SPS. You can include these widgets in your own Shiny apps without loading SPS main framework.</p>
</div>
</div>
<div id="extend-sps" class="section level1">
<h1 class="hasAnchor">
<a href="#extend-sps" class="anchor"></a>Extend SPS</h1>
<p>As a framework, SPS allows users to add/remove/customize individual tabs. At current version, you can only modify tabs in VS part(data tabs, plot tabs). Use <code>newTabData</code> to add a new data tab and <code>newTabPlot</code> for a plot tab. These functions will help you to <strong>create the tab file</strong>, store it to the <em>R</em> folder in your SPS project, and <strong>register tab information</strong> by adding records to <em>config/tabs.csv</em>. Most arguments are self-explanatory or can be easily understood by reading the the help file.</p>
</div>
<div id="how-to-deploy-the-application" class="section level1">
<h1 class="hasAnchor">
<a href="#how-to-deploy-the-application" class="anchor"></a>How to deploy the application</h1>
<p>There is a <em>deploy.R</em> file in the SPS project root and you can use it to deploy the app to <a href="https://www.shinyapps.io/">shinyapps.io</a>. You can use this file to set up your account and deploy the app. SPS is a deploy-ready framework. If you set up your account correctly, you can deploy it as soon as the project is created. It is recommended to create -&gt; test/run -&gt; deploy a Shiny app from Rstudio and follows <a href="https://shiny.rstudio.com/articles/shinyapps.html">this instruction</a>.</p>
</div>
<div id="version-information" class="section level1">
<h1 class="hasAnchor">
<a href="#version-information" class="anchor"></a>Version Information</h1>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<pre><code>## R Under development (unstable) (2020-08-07 r78979)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /home/dcassol/src/R-devel/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] systemPipeShiny_0.99.0 shinyTree_0.2.7        ggplot2_3.3.2         
## [4] stringr_1.4.0          shiny_1.5.0            magrittr_1.5          
## [7] glue_1.4.2             assertthat_0.2.1       BiocStyle_2.17.1      
## 
## loaded via a namespace (and not attached):
##  [1] fs_1.5.0                 bit64_4.0.5              httr_1.4.2              
##  [4] rprojroot_1.3-2          tools_4.1.0              backports_1.1.10        
##  [7] R6_2.4.1                 DT_0.15                  DBI_1.1.0               
## [10] lazyeval_0.2.2           colorspace_1.4-1         pushbar_0.1.0           
## [13] shinyjqui_0.3.3          withr_2.3.0              tidyselect_1.1.0        
## [16] bit_4.0.4                curl_4.3                 compiler_4.1.0          
## [19] formatR_1.7              shinyjs_2.0.0            desc_1.2.0              
## [22] plotly_4.9.2.1           rhandsontable_0.3.7      colourpicker_1.1.0      
## [25] bookdown_0.20            scales_1.1.1             askpass_1.1             
## [28] pkgdown_1.6.1            systemfonts_0.3.1        digest_0.6.25           
## [31] shinydashboardPlus_0.7.5 rmarkdown_2.3            pkgconfig_2.0.3         
## [34] htmltools_0.5.0          DOT_0.1                  styler_1.3.2            
## [37] fastmap_1.0.1            htmlwidgets_1.5.1        rlang_0.4.7             
## [40] rstudioapi_0.11          RSQLite_2.2.0            generics_0.0.2          
## [43] jsonlite_1.7.1           vroom_1.3.1.9000         dplyr_1.0.2             
## [46] Rcpp_1.0.5               munsell_0.5.0            lifecycle_0.2.0         
## [49] stringi_1.5.3            yaml_2.2.1               shinytoastr_2.1.1       
## [52] grid_4.1.0               blob_1.2.1               shinyFiles_0.8.0        
## [55] promises_1.1.1           shinydashboard_0.7.1     crayon_1.3.4            
## [58] miniUI_0.1.1.1           knitr_1.30               pillar_1.4.6            
## [61] igraph_1.2.5             markdown_1.1             codetools_0.2-16        
## [64] evaluate_0.14            V8_3.2.0                 data.table_1.13.0       
## [67] BiocManager_1.30.10      vctrs_0.3.4              httpuv_1.5.4            
## [70] networkD3_0.4            gtable_0.3.0             openssl_1.4.3           
## [73] purrr_0.3.4              tidyr_1.1.2              cpp11_0.2.1             
## [76] xfun_0.17                mime_0.9                 xtable_1.8-4            
## [79] rsvg_2.1                 later_1.1.0.1            ragg_0.3.1              
## [82] viridisLite_0.3.0        tibble_3.0.3             memoise_1.1.0           
## [85] shinyWidgets_0.5.3       ellipsis_0.3.1           shinyAce_0.4.1</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://lezhang.me/">Le Zhang</a>, <a href="https://www.dcassol.com/">Daniela Cassol</a>, <a href="https://github.com/mathrj">Ponmathi Ramasamy</a>, <a href="https://github.com/jianhaizhang">Jianhai Zhang</a>, <a href="https://github.com/gdmosher">Gordon Mosher</a>, <a href="http://girke.bioinformatics.ucr.edu/">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
