<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>
<span style="color:#5DA7D6;">s</span>ystem<span style="color:#5DA7D6;">P</span>ipe<span style="color:#5DA7D6;">S</span>hiny • systemPipeShiny</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="&lt;span style=" color:>
</head>
<body>
<p>system<span style="color:#5DA7D6;">P</span>ipe<span style="color:#5DA7D6;">S</span>hiny" /&gt;
<meta property="og:description" content="systemPipeShiny"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></p>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeShiny</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/systemPipeShiny.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/systemPipeR/systemPipeShiny/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="systemPipeShiny_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="systemPipeShiny_files/font-awesome-5.1.0/css/all.css" rel="stylesheet">
<link href="systemPipeShiny_files/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>
<span style="color:#5DA7D6;">s</span>ystem<span style="color:#5DA7D6;">P</span>ipe<span style="color:#5DA7D6;">S</span>hiny</h1>
                        <h4 class="author">Le Zhang</h4>
            <address class="author_afil">
      University of California Riverside<br><a class="author_email" href="mailto:#"></a><a href="mailto:le.zhang001@email.ucr.edu" class="email">le.zhang001@email.ucr.edu</a>
      </address>
                              <h4 class="author">Daniela Cassol</h4>
            <address class="author_afil">
      University of California Riverside<br><a class="author_email" href="mailto:#"></a><a href="mailto:danielac@ucr.edu" class="email">danielac@ucr.edu</a>
      </address>
                              <h4 class="author">Ponmathi Ramasamy</h4>
            <address class="author_afil">
      University of California Riverside<br><h4 class="author">Jianhai Zhang</h4>
            <address class="author_afil">
      University of California Riverside<br><h4 class="author">Gordon Mosher</h4>
            <address class="author_afil">
      University of California Riverside<br><h4 class="author">Thomas Girke</h4>
            <address class="author_afil">
      University of California Riverside<br><a class="author_email" href="mailto:#"></a><a href="mailto:tgirke@ucr.edu" class="email">tgirke@ucr.edu</a>
      </address>
                  
            <h4 class="date">Last update: 20 August, 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/systemPipeR/systemPipeShiny/blob/master/vignettes/systemPipeShiny.Rmd"><code>vignettes/systemPipeShiny.Rmd</code></a></small>
      <div class="hidden name"><code>systemPipeShiny.Rmd</code></div>

    </address>
</address>
</address>
</div>

    
    
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<style>/*Bootstrap colors*/
:root {
  --blue: #007bff;
  --indigo: #6610f2;
  --purple: #6f42c1;
  --pink: #e83e8c;
  --red: #dc3545;
  --orange: #fd7e14;
  --yellow: #ffc107;
  --green: #28a745;
  --teal: #20c997;
  --cyan: #17a2b8;
  --white: #fff;
  --gray: #6c757d;
  --gray-dark: #343a40;
  --primary: #337ab7;
  --secondary: #6c757d;
  --success: #28a745;
  --info: #17a2b8;
  --warning: #ffc107;
  --danger: #dc3545;
  --light: #f8f9fa;
  --dark: #343a40;
  --breakpoint-xs: 0;
  --breakpoint-sm: 576px;
  --breakpoint-md: 768px;
  --breakpoint-lg: 992px;
  --breakpoint-xl: 1200px;
  --font-family-sans-serif: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  --font-family-monospace: SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace;
}

/*
systemPipeShiny CSS Styles
*/

/*buttons, default is too ugly*/
.btn-default {
  color: #fff;
  background-color: var(--primary);
  border-color: #2e6da4;
}

.btn-default:hover{
  background-color: #4c92cf;
}

/*widget user in shinydashboardplus*/
.widget-user {
  overflow-y:auto;
}

/*for clearableTextInput component*/
.text-input-clearable {
  width: 100%;
  border:1px solid #ccc;
  border-radius: 3px;
  padding:1px 6px 1px 1px;
  display:inline-block;
}

.text-input-clearable input {
  width: 93%;
  border:none;
  background:none;
  outline:none;
  padding:0 0;
  margin:0 0;
  font:inherit;
}

.text-input-clearable span {
  float: right;
  height: 14px;
  font-size: 14px;
  cursor: pointer;
  color: #ccc;
  visibility:hidden;
}
.text-input-focused {
  border-color: #66afe9;
  outline: 0;
  -webkit-box-shadow: inset 0 1px 1px rgba(0,0,0,.075), 0 0 8px rgba(102, 175, 233, 1);
  -moz-box-shadow: inset 0 1px 1px rgba(0,0,0,.075), 0 0 8px rgba(102, 175, 233, 1);
  box-shadow: inset 0 1px 1px rgba(0,0,0,.075), 0 0 8px rgba(102, 175, 233, 1);
}

/*for gallery */
img-gallery{
  max-width:100%;
  height:auto;
  object-fit: cover;
  vertical-align: middle;
  text-align: center;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
}

img-gallery:hover {
  opacity: 0.7;
  transition: opacity 0.5s;
}


/*for sps hyperlink button  and tables*/
.href-button {
  background-color: var(--primary);
  border: none;
  border-radius: 10px 10px;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
  color: white;
  padding: 10px 25px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  font-size: 15px;
  margin: 4px 2px;
  cursor: pointer;
         }

.href-button:hover {
  opacity: 0.9;
  text-decoration: none;
  color: white;
  transition: opacity 0.5s;
}

.href-button:active {
  transform: translateY(2px);
}

.table-href td:first-child{
  width: 1px;
  vertical-align: middle;
  white-space: nowrap;
}

/* about tab background */
.widget-user-header {
    background: linear-gradient(-45deg, #428bca, #ccccff, #b3b3b3);
    background-size: 400% 400%;
    animation: gradient 10s ease infinite;
}

@keyframes gradient {
    0% {
        background-position: 0% 50%;
    }
    50% {
        background-position: 100% 50%;
    }
    100% {
        background-position: 0% 50%;
    }
}

/* sps description container*/
.desc-body.collapse {
  display: block;
  height: 150px;
  overflow: hidden;
}
.desc-body.in {
  height: auto;
}
.desc a.collapsed:after  {
    content: '+ Show More Description';
}
.desc a:not(.collapsed):after {
    content: '- Collapse Description';
}
.desc a {
  font-size: 1.1em;
}
.desc a:link {
  text-decoration: none;
}

/* sps control-panel*/
.control-panel {
    opacity: 0.7;
    transition: opacity .25s ease-in-out;
   -moz-transition: opacity .25s ease-in-out;
   -webkit-transition: opacity .25s ease-in-out;
}

.control-panel:hover { opacity: 1;}


/* sps hexagon icon and hex panel*/
.hex-container {
  display: inline-block;
  text-align: center;
  width: 100%;
}

.hex {
  opacity: 0.8;
}

.hex:hover {
  opacity: 1;
}

.powerby-link {
  font: italic bold 25px serif;
  fill: var(--primary);
  opacity: 0.5;
}

.powerby-link:hover {
  opacity: 1;
}

.hex-box {
  width: 3vw;
  min-width: 50px;
  display: inline-block;
  padding-top: 0;
}

.hex-container span {
 display: inline-block;
 text-align: center;
 font-size: 0.6vw;
}

.hex-panel {
  position: auto;
}
.hex-panel > .hex-item {
  display: inline-block;
  text-align: left;
  padding-left: 0;
  padding-right: 0;
  margin-left: 0;
  margin-right:0;
}

.hex-panel h5 {
  display: inline-block;
  margin-left: 20px;
  font-weight: 900;
  font-size: 1em;
  transform: translateY(-2.2vw);
}

/* sps canvas*/
.snap-drag {
  opacity: 0;
}

.snap-drag:hover {
  opacity: 1;
}

.snap-drag button{
  position: absolute;
  outline: none;
  background-color: Transparent;
  top: 2px;
  right: 4px;
}

.snap-drag button:focus {
  outline: 0 !important;
  box-shadow: none !important;
}

.snap-drag h4{
  margin-bottom: 0;
  padding-bottom: 10px;
}

a.bttn-default {
    background: #3c8dbc !important;
}

.sps-warning h4 {
    color: var(--danger);
}
.sps-warning h4::before{
    content: "\2022 ";
}
</style>
<p><img src="../inst/app/www/img/sps.png" width="70%" style="display: block; margin: auto;"></p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><strong><span style="color:#5DA7D6;">s</span>ystem<span style="color:#5DA7D6;">P</span>ipe<span style="color:#5DA7D6;">S</span>hiny</strong>(SPS) is a <a href="https://shiny.rstudio.com/">Shiny</a>-based R/Bioconductor package that has two main purposes:</p>
<ol style="list-style-type: decimal">
<li>To help non-programmers like wet lab researchers on data analysis workflow management and downstream data visualizations with a graphical user interface (UI).</li>
<li>An user friendly framework for both beginners and advanced developers to fast extend with more visualization options(more plot types) and to quickly deploy the app to a Shiny sever.</li>
</ol>
<p>You can try our demo at <a href="https://tgirke.shinyapps.io/systemPipeShiny/">here</a>. There are 5 different interactive tutorials embedded on the demo that you can follow(right upper corner). Each tutorial covers a major topic of SPS and show how you can interactive with app.</p>
<p>This vignette introduces main features and functionalities of SPS. Due to size limit, a full version of users’ manual will be later updated at <a href="https://systempipe.org/systemPipeShiny-book/">SytemPipeShiny-book</a> which will include more small demos, screenshots, and <em>gif</em> illustrations.</p>
<div id="install" class="section level2">
<h2 class="hasAnchor">
<a href="#install" class="anchor"></a>Install</h2>
<p>To install SPS:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="kw">if</span> (!<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="kw">quietly</span> <span class="kw">=</span> <span class="fl">TRUE</span>)) <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"systemPipeR/systemPipeShiny"</span>, <span class="kw">build_vignettes</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
    <span class="kw">dependencies</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p>If you are on Linux, you also need(use Ubuntu as example):</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">sudo</span> apt-get install libv8-dev</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="fu">sudo</span> apt-get install libssl-dev</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="fu">sudo</span> apt-get install libcurl4-openssl-dev</span></code></pre></div>
<p><em>libcurl4-openssl-dev</em> may not be required for Ubuntu &gt;= 18.04.</p>
<p>Depending on the Linux distributions, command can be slightly different.</p>
</div>
<div id="quick-start" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h2>
<p>Before starting with SPS, you need to create a SPS project:</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">sps_tmp_dir</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span>()
<span class="fu"><a href="../reference/spsInit.html">spsInit</a></span>(<span class="kw">dir_path</span> <span class="kw">=</span> <span class="no">sps_tmp_dir</span>, <span class="kw">change_wd</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p>Here for building the vignette we are not switching to the app directory and are using a temp directory, in a real case, you <strong>shouldn’t</strong> store you SPS project in a <strong>temp directory</strong>.</p>
<p>By default, you will get a folder created under your current working directory, the name is <code>SPS_</code> + <code>today's date</code> in <em>ISO 8601</em> format, e.g.  SPS_20200820</p>
<p style="color:var(--info)">
Note: all SPS date or time stamp objects use <em>ISO 8601</em> format
</p>
<p>By default, your working directory should be set inside that project folder automatically. If you are using Rstudio, three main files will be opened for you: <code>global.R</code>, <code>ui.R</code> and <code>server.R</code>. Now you can just run the app by type <code><a href="https://shiny.rstudio.com/reference/shiny/latest/runApp.html">shiny::runApp()</a></code> in console or click on the <img src="img/app_start.png" style="display: inline; padding-left:0; width:10px; height:15px;"><code>Run App</code> button on top right corner of the any these 3 files in Rstudio. In your global.R, scroll down to the bottom, you should see:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">sps_app</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/sps.html">sps</a></span>(<span class="kw">vstabs</span> <span class="kw">=</span> <span class="st">""</span>, <span class="kw">server_expr</span> <span class="kw">=</span> {
    <span class="fu"><a href="../reference/msg.html">msg</a></span>(<span class="st">"Custom expression runs -- Hello World"</span>, <span class="st">"GREETING"</span>,
        <span class="st">"green"</span>)
})</pre></body></html></div>
<p>This is the SPS main function. You can load/unload tabs by providing tab IDs in <code>vstabs</code> argument, like <code>c("tab1", "tab2)</code>. See <code>config/tabs.csv</code> in your project folder for what tabs IDs can be load and other tab information. Currently you can only load/unload visualization tabs, type of <em>vs</em>; core tabs(<em>core</em>) and workflow tabs (<em>wf</em>) are loaded automatically.</p>
</div>
</div>
<div id="sps-structure" class="section level1">
<h1 class="hasAnchor">
<a href="#sps-structure" class="anchor"></a>SPS structure</h1>
<div id="app-directory-structure" class="section level2">
<h2 class="hasAnchor">
<a href="#app-directory-structure" class="anchor"></a>App directory structure</h2>
When a SPS project is created, you would expect the directory contains:
<pre>
SPS_xx/  
├── server.R               <span class="text-success">|</span> 
├── global.R               <span class="text-success">| Most important server, UI and global files, unless special needs, `global.R` is the only file you need to edit manually</span>   
├── ui.R                   <span class="text-success">|</span>  
├── deploy.R               <span class="text-info">| Deploy helper file</span>  
├── config                 <span class="text-success">| Important app config files. Do not edit them if you don't know</span>  
│   ├── sps.db             
│   ├── sps_options.yaml   
│   └── tabs.csv           
├── data                   <span class="text-primary">| App example data files</span> 
│   ├── xx.csv             
├── R                      <span class="text-primary">| All SPS additional tab files and helper R function files</span> 
│   ├── tab_xx.R            
├── README.md              
├── results                <span class="text-gray">| No in use for this current version, you can store some data been generated from the app</span> 
│   └── README.md          
└── www                    <span class="text-primary">| Internet resources</span>  
    ├── about              <span class="text-gray">| About tab information</span> 
    │   └── xx.md          
    ├── css                <span class="text-info">| CSS files</span>  
    │   └── sps.css         
    ├── img                <span class="text-info">| App image resources</span>    
    │   └── xx.png         
    ├── js                 <span class="text-info">| Javascripts</span>
    │   └── xx.js           
    ├── loading_themes     <span class="text-info">| Loading screen files</span> 
    │   └── xx.html         
    └── plot_list          <span class="text-info">| Image files for plot gallery</span>  
        └── plot_xx.jpg      
</pre>
</div>
<div id="app-ui-structure" class="section level2">
<h2 class="hasAnchor">
<a href="#app-ui-structure" class="anchor"></a>App UI structure</h2>
<p>SPS provides two core components – workflow management and data visualization. The UI is designed as a dashboard(<em>Fig 1</em>). Individual components are classified into two categories of tabs (<em>Fig 1. a-b</em>).</p>
<center>
<img src="img/sps_ui_main.jpg"><strong>Figure 1. SPS UI</strong>
</center>
<p>The app UI structure can be understand as following(<em>Fig 2</em>). Users first need to choose either the workflow or visualization part to start with. Workflow part helps users to prepare data analysis workflow files in <a href="http://systempipe.org/">systemPipeR</a>(SPR) format. Visualization part helps users to make different types of plots, take interactive snapshot of plots and combine/compare different plots on the “Canvas” workbench.</p>
<center>
<img src="img/sps_structure.png"><strong>Figure 2. SPS Structure</strong>
</center>
</div>
</div>
<div id="ui-details" class="section level1">
<h1 class="hasAnchor">
<a href="#ui-details" class="anchor"></a>UI details</h1>
<div id="workflow-management" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-management" class="anchor"></a>Workflow management</h2>
<p>Files can be prepared for a data analysis workflow including the metadata(targets) file, workflow step file(Rmarkdown formatted) and a workflow config file. Each of these files are prepared in a SPS tab. You can find them at the left side navigation bar. SPS helps users to check the formats and requirements of each file. Users can click on the <button type="button" class="btn btn-default"> <i class="fa fa-paper-plane"></i> Add to task </button> button on the page to check for problems. If everything is correct, a pop-up messge will show to users or if anything is wrong, a detailed window will pop up to indicate errors.</p>
<p>In the upcomming version of SPS, users can directly run a workflow on the app, but for now users need to download these files and copy them to a SPR project and run from command line. You can download these files by clicking <button class="btn btn-default"> <i class="fa fa-download"></i> Save </button>.</p>
<div id="plot_template-pg-pg-container" class="tab-pane" style="padding-left: 130px;">
<div class="control-panel draggable ui-draggable ui-draggable-handle" style="width: 400px; height: auto; background-color: var(--light);">
<div class="row">
<div class="col-sm-1">

</div>
<div class="col-sm-9">
<h5>
Workflow-prepare
</h5>
</div>
<div class="col-sm-2">
<button class="btn btn-default" data-target="#pg-example" data-toggle="collapse" aria-expanded="true">
<i class="fa fa-minus"></i>
</button>
</div>
</div>
<div id="pg-example" class="in collapse" aria-expanded="true" style="">
<p><img src="img/sps_progress_bar.png" style="width:200% !important; padding:0; margin:0"></p>
</div>
</div>
</div>
<p>A typical workflow tab will look like this one(<em>Fig3</em>).</p>
<center>
<img src="img/sps_ui_targets.png" alt="systemPipeShiny"><strong>Figure 3. Workflow Management</strong>
</center>
</div>
<div id="data-visualization" class="section level2">
<h2 class="hasAnchor">
<a href="#data-visualization" class="anchor"></a>Data Visualization</h2>
<div id="tabs" class="section level3">
<h3 class="hasAnchor">
<a href="#tabs" class="anchor"></a>Tabs</h3>
<p>There are two types of data visualization(VS) tabs: <em>data tab</em> and <em>plot tab</em>. A data tab is where you prepare plotting data, including different preprocess options. Depending on the data tab and preprocess options you choose, you will be given different plot tab options.</p>
<p>You should first going to the VS main tab by clicking <em><i class="fa fa-images"></i>Visualization</em> and read the instructions there.</p>
<p>There are two ways to make a plot:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Begin with a data tab</strong>: If you know have run some data analysis workflows and have obtained results but don’t know what you can plot from these results, browse the table on visualization main tab. The table is sorted by workflows as rows. Based on the data type you have, you can click on the data type in the table, you will jump to the corresponding data preparation tab and you will see what plots you can make after data preparation.</li>
<li>
<strong>Begin with a plot tab</strong>: If you already know what plots you want, there is a gallery on visualization main tab. You can click on the plot you want and it will jump to the plot tab directly. You will see in order to make this plot what input data need to be prepared. By choosing under the <strong>Data Preparation</strong>, you will go to a data tab and once you have prepared the required data, you will be guided back to this plotting tab.</li>
</ol>
<p>All VS tabs has a start-up button <button type="button" class="btn btn-default">Start with this tab</button> which will first check for packages requirements. If you don’t have the required package for the tab, there will be a pop-up window telling you how to install the required packages. In addition to package checks, plot tabs will also check for required plotting data. Depending on missing or unqualified data, you will see different error messages on the bottom right corner. See <a href="#dual-end-logging">SPS logging</a> in for more information. If everything is good, you will still see a green message and the main tab UI will show up and you can continue.</p>
</div>
<div id="data-tabs" class="section level3">
<h3 class="hasAnchor">
<a href="#data-tabs" class="anchor"></a>Data tabs</h3>
<p>Once you have passed the package checks, you should see the data tab main UI. Please read the instructions on the top of each tab for how to prepare your data and data format requirements. All SPS data tabs comes with an example data set, and if you are not sure what data to prepare, you can first switch to the example data set and play with it. When you are familiar with the tab, you can upload your own data set.</p>

<div id="df_template-data_source" class="radioGroupButtons shiny-bound-input" style="padding-left: 60px;">
  <div aria-label="..." class="btn-group btn-group-justified btn-group-container-sw" data-toggle="buttons" role="group">
    <div class="btn-group btn-group-toggle" role="group">
      <button class="btn radiobtn btn-default active">
        <input type="radio" autocomplete="off" name="df_template-data_source" value="upload" checked>
        Upload
      </button>
    </div>
    <div class="btn-group btn-group-toggle" role="group">
      <button class="btn radiobtn btn-default">
        <input type="radio" autocomplete="off" name="df_template-data_source" value="eg">
        Example
      </button>
    </div>
  </div>
</div>    

<p style="color:var(--info)">
Note: Upload file selection can be different depending on the mode, see <a href="#app-options">App options</a>
</p>

<div class="input-group" style="width: 30%; padding-left: 60px">
  <label class="input-group-btn input-group-prepend">
    <span class="btn btn-file btn-default">
      Browse
      <input id="df_template-file_upload" name="df_template-file_upload" type="file" style="display: none;"></span>
  </label>
  <input type="text" class="form-control" placeholder="No file selected" readonly>
</div>

</div>
<div id="plot-tabs" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-tabs" class="anchor"></a>Plot tabs</h3>
</div>
</div>
<div id="snapshot-and-canvas" class="section level2">
<h2 class="hasAnchor">
<a href="#snapshot-and-canvas" class="anchor"></a>Snapshot and Canvas</h2>
<center>
<img src="img/sps_ui_canvas.png" alt="systemPipeShiny"><strong>Figure 4. Canvas</strong>
</center>
</div>
</div>
<div id="configuration" class="section level1">
<h1 class="hasAnchor">
<a href="#configuration" class="anchor"></a>Configuration</h1>
</div>
<div id="features" class="section level1">
<h1 class="hasAnchor">
<a href="#features" class="anchor"></a>Features</h1>
<div id="user-friendly" class="section level2">
<h2 class="hasAnchor">
<a href="#user-friendly" class="anchor"></a>User friendly</h2>
<p>SPS is easy to use for people different backgrounds. The GUI operations require no programming experience. Detailed design templates are provided for advanced users allowing them to add new features(tabs) to SPS.</p>
</div>
<div id="dual-end-logging" class="section level2">
<h2 class="hasAnchor">
<a href="#dual-end-logging" class="anchor"></a>Dual-end logging</h2>
<p>Messages, warnings and errors from R functions are automatically captured and logged on both the server and client ends. The corresponding information can be displayed in a message box.</p>
</div>
<div id="progress-tracking" class="section level2">
<h2 class="hasAnchor">
<a href="#progress-tracking" class="anchor"></a>Progress tracking</h2>
<p>A progress tracking panel is provided on each tab to notify users about run status.</p>
</div>
<div id="app-options" class="section level2">
<h2 class="hasAnchor">
<a href="#app-options" class="anchor"></a>App options</h2>
<p>a robust exception handling system has been implemented (similar to Shiny options), that provides error solutions to to users, e.g. invalid parameter settings. - upload options</p>
</div>
<div id="modular-isolation" class="section level2">
<h2 class="hasAnchor">
<a href="#modular-isolation" class="anchor"></a>Modular isolation</h2>
<p>SPS is built on Shiny modules, which provides local scope isolation between each tab. Objects on one tab do not conflict with other tabs. They work like individual posts in a forum. To enable cross-tab communication, SPS also supports global scope interactions. ## A variety of additional shiny widgets: There are many useful HTML components that can be use outside out SPS. You can include these widgets in your own Shiny apps without loading SPS main framework.</p>
</div>
<div id="canvas" class="section level2">
<h2 class="hasAnchor">
<a href="#canvas" class="anchor"></a>Canvas</h2>
<p>Under this workbench users can take snapshots of different plots, and combine or resize them. This feature is useful for generating complex scientific summary graphics.</p>
<div id="systempipeshiny" class="section level3">
<h3 class="hasAnchor">
<a href="#systempipeshiny" class="anchor"></a>systemPipeShiny</h3>
<p>Shiny Web Interface Shiny application for interactive workflow management and results visualization generated by <em>systemPipeR</em></p>
</div>
</div>
</div>
<div id="setting-up-the-data" class="section level1">
<h1 class="hasAnchor">
<a href="#setting-up-the-data" class="anchor"></a>Setting up the data</h1>
</div>
<div id="launching-the-application" class="section level1">
<h1 class="hasAnchor">
<a href="#launching-the-application" class="anchor"></a>Launching the application</h1>
</div>
<div id="description-of-the-user-interface" class="section level1">
<h1 class="hasAnchor">
<a href="#description-of-the-user-interface" class="anchor"></a>Description of the user interface</h1>
<div id="dashboard" class="section level2">
<h2 class="hasAnchor">
<a href="#dashboard" class="anchor"></a>dashboard</h2>
</div>
<div id="layout" class="section level2">
<h2 class="hasAnchor">
<a href="#layout" class="anchor"></a>layout</h2>
</div>
<div id="each-tab" class="section level2">
<h2 class="hasAnchor">
<a href="#each-tab" class="anchor"></a>each tab</h2>
</div>
<div id="workflow-management-system" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-management-system" class="anchor"></a>Workflow Management system</h2>
</div>
</div>
<div id="description-of-systempipeshiny-functionality" class="section level1">
<h1 class="hasAnchor">
<a href="#description-of-systempipeshiny-functionality" class="anchor"></a>Description of <em>systemPipeShiny</em> functionality</h1>
</div>
<div id="sps-configurations" class="section level1">
<h1 class="hasAnchor">
<a href="#sps-configurations" class="anchor"></a>SPS configurations</h1>
<div id="the-config-folder" class="section level2">
<h2 class="hasAnchor">
<a href="#the-config-folder" class="anchor"></a>the config folder</h2>
</div>
<div id="sps-options" class="section level2">
<h2 class="hasAnchor">
<a href="#sps-options" class="anchor"></a>SPS options</h2>
</div>
<div id="sps-tab" class="section level2">
<h2 class="hasAnchor">
<a href="#sps-tab" class="anchor"></a>SPS tab</h2>
</div>
<div id="sps-database" class="section level2">
<h2 class="hasAnchor">
<a href="#sps-database" class="anchor"></a>SPS database</h2>
</div>
</div>
<div id="for-developers" class="section level1">
<h1 class="hasAnchor">
<a href="#for-developers" class="anchor"></a>For developers</h1>
<div id="tab-control" class="section level2">
<h2 class="hasAnchor">
<a href="#tab-control" class="anchor"></a>Tab control</h2>
<div id="tab-classification" class="section level3">
<h3 class="hasAnchor">
<a href="#tab-classification" class="anchor"></a>Tab classification</h3>
</div>
<div id="new-data-tab" class="section level3">
<h3 class="hasAnchor">
<a href="#new-data-tab" class="anchor"></a>New data tab</h3>
</div>
<div id="new-plot-tab" class="section level3">
<h3 class="hasAnchor">
<a href="#new-plot-tab" class="anchor"></a>New plot tab</h3>
</div>
<div id="load-new-tab-to-the-framework" class="section level3">
<h3 class="hasAnchor">
<a href="#load-new-tab-to-the-framework" class="anchor"></a>Load new tab to the framework</h3>
</div>
<div id="remove-a-tab" class="section level3">
<h3 class="hasAnchor">
<a href="#remove-a-tab" class="anchor"></a>Remove a tab</h3>
</div>
<div id="a-empty-tab-for-advanced-devleopers" class="section level3">
<h3 class="hasAnchor">
<a href="#a-empty-tab-for-advanced-devleopers" class="anchor"></a>a Empty tab for advanced devleopers</h3>
</div>
</div>
<div id="section" class="section level2">
<h2></h2>
</div>
<div id="visualization" class="section level2">
<h2 class="hasAnchor">
<a href="#visualization" class="anchor"></a>Visualization</h2>
<p><em>systemPipeShiny</em> offers interactive data visualization to users and supports graphics in workflow reports. In systemPipeShiny’s visualization tab, there are many features for data visualization of various statistical results. Users can upload different input data types, and apply various options for preprocessing those datasets. Users can then create downstream analysis plots, as per the type of uploaded data. Some available plotting options include: bar plots of differentially expressed genes, heat maps, dendrogram, principal component analysis (PCA) plots, and multidimensional scaling (MDS) plots. Depending on the nature of the plots, there are also options to adjust the plot such as normalizing the data. Additionally, systemPipeShiny provides users with plot templates and plotting functions that they can then customize according to their necessities for visualization.</p>
<p>Table with all exported functions</p>
<table class="table">
<colgroup>
<col width="20%">
<col width="79%">
</colgroup>
<thead><tr class="header">
<th>Function Name</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>exploreDSS</code></td>
<td>Transform raw read counts using the  package</td>
</tr>
<tr class="even">
<td><code>plotExploreDSS</code></td>
<td>Scatterplot of transformed counts from two samples</td>
</tr>
<tr class="odd">
<td><code>plotPCA_sps</code></td>
<td>Plots PCA from a count matrix</td>
</tr>
<tr class="even">
<td><code>plotMDS</code></td>
<td>Plots MDS from a count matrix</td>
</tr>
<tr class="odd">
<td><code>plotTSNE</code></td>
<td>Plots t-SNE from a count matrix</td>
</tr>
<tr class="even">
<td><code>plotGLM</code></td>
<td>Plots GLM-PCA from a count matrix</td>
</tr>
<tr class="odd">
<td><code>plotHeatMap</code></td>
<td>Plots Heatmap from a count matrix</td>
</tr>
<tr class="even">
<td><code>plot_MA</code></td>
<td>Plots a clustering dendrogram from a count matrix</td>
</tr>
<tr class="odd">
<td><code>run_volcano</code></td>
<td>Plots a Volcano Plot from an edgeR or deseq2 dataframe</td>
</tr>
<tr class="even">
<td><code>deg_edgeR</code></td>
<td>Plots barplot of DEGs from a count matrix, returns an edgeR DF</td>
</tr>
<tr class="odd">
<td><code>deg_deseq2</code></td>
<td>Plots barplot of DEGs from a count matrix, returns an deseq2 DF</td>
</tr>
</tbody>
</table>
<div id="data-transformations-and-visualization" class="section level3">
<h3 class="hasAnchor">
<a href="#data-transformations-and-visualization" class="anchor"></a>Data transformations and visualization</h3>
<p>To show the effect of the transformation, in the figure below we plot the first sample against the second, first simply using the <code>log2</code> function, and then using the <code>VST</code> and <code>rlog-transformed</code> values. For the log2 approach, we need to first estimate size factors to account for sequencing depth, and then specify <code>normalized=TRUE</code>. Sequencing depth correction is done automatically for the <code>vst</code> and <code>rlog.</code></p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">targetspath</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"systemPipeR"</span>)
<span class="no">targets</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span>(<span class="no">targetspath</span>, <span class="kw">comment</span> <span class="kw">=</span> <span class="st">"#"</span>)
<span class="no">cmp</span> <span class="kw">&lt;-</span> <span class="kw pkg">systemPipeR</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/readComp.html">readComp</a></span>(<span class="kw">file</span> <span class="kw">=</span> <span class="no">targetspath</span>, <span class="kw">format</span> <span class="kw">=</span> <span class="st">"matrix"</span>,
    <span class="kw">delim</span> <span class="kw">=</span> <span class="st">"-"</span>)
<span class="no">countMatrixPath</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"countDFeByg.xls"</span>,
    <span class="kw">package</span> <span class="kw">=</span> <span class="st">"systemPipeR"</span>)
<span class="no">countMatrix</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span>(<span class="no">countMatrixPath</span>, <span class="kw">row.names</span> <span class="kw">=</span> <span class="fl">1</span>)
<span class="fu">plot.exploreDDS</span>(<span class="no">countMatrix</span>, <span class="no">targets</span>, <span class="kw">cmp</span> <span class="kw">=</span> <span class="no">cmp</span><span class="kw">[[</span><span class="fl">1</span>]], <span class="kw">preFilter</span> <span class="kw">=</span> <span class="fl">10</span>,
    <span class="kw">sample</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">3</span>, <span class="fl">4</span>))</pre></body></html></div>
</div>
</div>
<div id="barplot" class="section level2">
<h2 class="hasAnchor">
<a href="#barplot" class="anchor"></a>Barplot</h2>
<p>A barplot for analysis of differentially expressed genes (DEGs) can be plotted using functions <code>deg_edgeR</code> or <code>deg_deseq2</code>. The function <code>deg_edgeR</code> uses the <code>edgeR</code> package <span class="citation">(<span class="citeproc-not-found" data-reference-id="Robinson2010-uk"><strong>???</strong></span>)</span> to create an <code>edgeR</code> data frame. Alternatively, the function <code>deg_deseq2</code> uses the <code>DESeq2</code> package <span class="citation">(<span class="citeproc-not-found" data-reference-id="Love2014-sh"><strong>???</strong></span>)</span> to create an <code>DESeq2</code> data frame. Using the <code>filterDEGs</code> function, it filters and plots DEG results for up and down regulated genes in a barplot.</p>
</div>
<div id="heatmap" class="section level2">
<h2 class="hasAnchor">
<a href="#heatmap" class="anchor"></a>Heatmap</h2>
<p>A heatmap of the results of hierarchical clustering performed with the <code>hclust</code> function can be created with the <code>run_HEAT</code> function. The sample-wise Spearman correlation coefficients are computed before hierarchical clustering. The count data frame can be transformed with the <code>rlog</code> or Variance-stabilizing Transformation (<code>vst</code>) methods from the <code>DESeq2</code> package, or can be done without transformation.</p>
</div>
<div id="dendrogram" class="section level2">
<h2 class="hasAnchor">
<a href="#dendrogram" class="anchor"></a>Dendrogram</h2>
<p>A dendrogram of the results of hierarchical clustering performed with the <code>hclust</code> function can be created with the <code>run_CLUST</code> function. The sample-wise Spearman correlation coefficients are computed, and then the results are transformed to a distance matrix before the hierarchical clustering is performed. The count data frame can be transformed with the <code>rlog</code> or Variance-stabilizing Transformation (<code>vst</code>) methods from the <code>DESeq2</code> package, or can be done without transformation.</p>
</div>
<div id="t-sne-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#t-sne-plot" class="anchor"></a>t-SNE plot</h2>
<p>A Barnes-Hut t-Distributed Stochastic Neighbor Embedding (t-SNE) plot can be created using the <code>run_TSNE</code> function, which uses the <code>Rtsne</code> package <span class="citation">(<span class="citeproc-not-found" data-reference-id="Krijthe2015"><strong>???</strong></span>)</span> to compute t-SNE values. The function removes duplicates in the input data frame, sets a seed for reproducility, performs an initial PCA step. The function also allows for a user-set perplexity value for the computation.</p>
</div>
<div id="pca-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#pca-plot" class="anchor"></a>PCA plot</h2>
<p>A Principal Component Analysis (PCA) plot can be created using the <code>run_PCA</code> function which uses the <code>DESeq2</code> package. The input data frame can be transformed with the <code>rlog</code> or Variance-stabilizing Transformation (<code>vst</code>) methods from the <code>DESeq2</code> package, or can be done without transformation. In addition, generalized principal component analysis (GLM-PCA) for dimension reduction of non-normally distributed data can be plotted with the <code>run_GLM</code> function <span class="citation">(<span class="citeproc-not-found" data-reference-id="Townes2019"><strong>???</strong></span>)</span>. This option does not offer transformation or normalization of raw data.</p>
</div>
<div id="mds-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#mds-plot" class="anchor"></a>MDS plot</h2>
<p>A Multidimensional Scaling (MDS) plot can be created using the <code>run_MDS</code> function. The input data frame can be transformed with either the <code>rlog</code> or Variance-stabilizing Transformation (<code>vst</code>) methods from the <code>DESeq2</code> package. From the input data, it computes a spearman correlation-based distance matrix and performs MDS analysis on it.</p>
</div>
<div id="volcano-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#volcano-plot" class="anchor"></a>Volcano plot</h2>
<p>A volcano plot of DEGs data frame can be plotted using the function <code>run_volcano</code>. Using the resulting data frame from <code>run_edgeR</code> or <code>run_deseq2</code>, the function plots a volcano plot using False Discovery Rate and Log Fold Change thresholds for the sample comparison specified by the user.</p>
</div>
</div>
<div id="how-to-use-this-function-on-the-shiny-application" class="section level1">
<h1 class="hasAnchor">
<a href="#how-to-use-this-function-on-the-shiny-application" class="anchor"></a>How to use this function on the Shiny application</h1>
</div>
<div id="how-to-custom-the-visualization-tab" class="section level1">
<h1 class="hasAnchor">
<a href="#how-to-custom-the-visualization-tab" class="anchor"></a>How to custom the Visualization Tab</h1>
</div>
<div id="for-developers-1" class="section level1">
<h1 class="hasAnchor">
<a href="#for-developers-1" class="anchor"></a>For developers</h1>
<div id="coventions-for-this-app" class="section level2">
<h2 class="hasAnchor">
<a href="#coventions-for-this-app" class="anchor"></a>Coventions for this app</h2>
<div id="app-structure" class="section level3">
<h3 class="hasAnchor">
<a href="#app-structure" class="anchor"></a>1. App structure</h3>
<ol style="list-style-type: decimal">
<li>Directories:
<ol style="list-style-type: decimal">
<li>config: config files</li>
<li>data: example datasets</li>
<li>results: not in use for now but users can store data generated from the app</li>
<li>R: all functions, tabs. This folder will be automatically sourced</li>
<li>www: html web resources, will be treated as root of frontend resources</li>
</ol>
</li>
</ol>
<p>Each folder has a README.md. You can know more information by reading that file.</p>
</div>
<div id="important-files" class="section level3">
<h3 class="hasAnchor">
<a href="#important-files" class="anchor"></a>2. Important files</h3>
<ol style="list-style-type: decimal">
<li>Three big files in the main directory:
<ul>
<li>
<strong>global.R</strong>: all sourcing, load library, global variables.</li>
<li>
<strong>server.R</strong>: top level server function, automatically generated by <code><a href="../reference/sps.html">sps()</a></code>.</li>
<li>
<strong>ui.R</strong>: top level UI, automatically generated by <code><a href="../reference/sps.html">sps()</a></code>.</li>
</ul>
</li>
</ol>
<p>All you need is to change code in <code>global</code></p>
</div>
<div id="naming" class="section level3">
<h3 class="hasAnchor">
<a href="#naming" class="anchor"></a>3. Naming</h3>
<ol style="list-style-type: decimal">
<li>tabs:
<ul>
<li>All store in <code>R</code> folder;</li>
<li>All should be named as <code>tab_xx.R</code>; if a submodule contains submodule. If this tab is a sub tab, name it <code>tab_type_id.R</code>, e.g. a tab for visualization new data type will be <code>tab_vs_data_xxx.R</code>, a new plot tab will be <code>tab_vs_plot_xxx.R</code>.</li>
<li>All tab info should also be updated in <code>tabs.csv</code> as the tab metadata.
<ul>
<li>visualization data tabs should hava tab name as <code>data_xx</code>, plot tab should be <code>plot_xx</code>
</li>
</ul>
</li>
</ul>
</li>
<li>functions:
<ul>
<li>in each <em>tab</em> file, there should be one <code>UI</code> function and one <code>server</code> function and give both functions and name space the same ID as the file name: e.g. a file named <code>tab_sub1.R</code>, UI function will be <code>sub1UI</code>, server will be <code>sub1Server</code> and in top level UI and server call them <code>sub1UI("sub1", ...)</code>, <code>callModule(tab1Server, "tab1", ...)</code>.</li>
</ul>
</li>
</ol>
</div>
</div>
<div id="standard-for-visualization" class="section level2">
<h2 class="hasAnchor">
<a href="#standard-for-visualization" class="anchor"></a>4. Standard for visualization</h2>
<ol style="list-style-type: decimal">
<li>Plots
<ul>
<li>In princple, datasets should be plotting ready (no need to preprocess data).</li>
<li>For some plots that are very specific to some workflows, simple preprocess is okay.</li>
<li>Always use a button to update (re-plot) the graph, realtime rendering can be expensive.</li>
</ul>
</li>
</ol>
</div>
<div id="objects-saved-in-shared" class="section level2">
<h2 class="hasAnchor">
<a href="#objects-saved-in-shared" class="anchor"></a>5. Objects saved in <code>shared</code>
</h2>
<p>“Shared” is first defined in the top level server as a shiny <code>ReactiveValues</code> object to hold data that can be passed around tabs(modules). This is very important if you want to transfer like a dataframe from df tabs to plotting tabs.</p>
<p>To access values in <code>shared</code>, use <code><a href="https://rdrr.io/r/base/Extract.html">$</a></code> or <code>[['name']]</code>, e.g. <code>shared$xxx$subxxx</code> or <code>shared[['xxx']][['subxxx']]</code>.</p>
<div id="default-stored-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#default-stored-objects" class="anchor"></a>default stored objects</h3>
<ul>
<li>wf_flags: bool values to indicate required files status for the workflow
<ul>
<li>targets_ready, wf_ready, wf_conf_ready: target file, workflow Rmd file, config yaml file</li>
</ul>
</li>
<li>targets: targets file
<ul>
<li>df: dataframe, which will be used to display from top push bar</li>
<li>file: string, the temp path of edited targets. When <code>add to task</code> is clicked and check passed, this file will be write to temp with <strong>target header</strong>
</li>
</ul>
</li>
<li>count: count table
<ul>
<li>df, file: same as targets</li>
</ul>
</li>
<li>config: configuration yaml file
<ul>
<li>file: same as targets</li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div id="how-to-deploy-the-application" class="section level1">
<h1 class="hasAnchor">
<a href="#how-to-deploy-the-application" class="anchor"></a>How to deploy the application</h1>
</div>
<div id="version-information" class="section level1">
<h1 class="hasAnchor">
<a href="#version-information" class="anchor"></a>Version Information</h1>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></body></html></div>
<pre><code>## R Under development (unstable) (2020-08-07 r78979)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /home/dcassol/src/R-devel/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets 
## [6] methods   base     
## 
## other attached packages:
## [1] systemPipeShiny_0.99.0 shinyTree_0.2.7       
## [3] ggplot2_3.3.2          stringr_1.4.0         
## [5] shiny_1.5.0            magrittr_1.5          
## [7] glue_1.4.1             assertthat_0.2.1      
## [9] BiocStyle_2.17.0      
## 
## loaded via a namespace (and not attached):
##   [1] colorspace_1.4-1           
##   [2] ellipsis_0.3.1             
##   [3] rprojroot_1.3-2            
##   [4] markdown_1.1               
##   [5] XVector_0.29.3             
##   [6] GenomicRanges_1.41.6       
##   [7] fs_1.5.0                   
##   [8] rstudioapi_0.11            
##   [9] DT_0.15                    
##  [10] bit64_4.0.2                
##  [11] AnnotationDbi_1.51.3       
##  [12] codetools_0.2-16           
##  [13] splines_4.1.0              
##  [14] geneplotter_1.67.0         
##  [15] knitr_1.29                 
##  [16] jsonlite_1.7.0             
##  [17] annotate_1.67.1            
##  [18] shinydashboard_0.7.1       
##  [19] BiocManager_1.30.10        
##  [20] compiler_4.1.0             
##  [21] httr_1.4.2                 
##  [22] backports_1.1.8            
##  [23] Matrix_1.2-18              
##  [24] fastmap_1.0.1              
##  [25] lazyeval_0.2.2             
##  [26] formatR_1.7                
##  [27] later_1.1.0.1              
##  [28] htmltools_0.5.0            
##  [29] tools_4.1.0                
##  [30] igraph_1.2.5               
##  [31] gtable_0.3.0               
##  [32] GenomeInfoDbData_1.2.3     
##  [33] dplyr_1.0.2                
##  [34] rsvg_2.1                   
##  [35] V8_3.2.0                   
##  [36] Rcpp_1.0.5                 
##  [37] Biobase_2.49.0             
##  [38] styler_1.3.2               
##  [39] pkgdown_1.5.1              
##  [40] pushbar_0.1.0              
##  [41] vctrs_0.3.2                
##  [42] shinyjqui_0.3.3            
##  [43] xfun_0.16                  
##  [44] networkD3_0.4              
##  [45] mime_0.9                   
##  [46] miniUI_0.1.1.1             
##  [47] lifecycle_0.2.0            
##  [48] XML_3.99-0.5               
##  [49] shinyAce_0.4.1             
##  [50] zlibbioc_1.35.0            
##  [51] MASS_7.3-52                
##  [52] scales_1.1.1               
##  [53] vroom_1.3.0                
##  [54] colourpicker_1.0           
##  [55] promises_1.1.1             
##  [56] parallel_4.1.0             
##  [57] SummarizedExperiment_1.19.6
##  [58] RColorBrewer_1.1-2         
##  [59] yaml_2.2.1                 
##  [60] curl_4.3                   
##  [61] memoise_1.1.0              
##  [62] stringi_1.4.6              
##  [63] RSQLite_2.2.0              
##  [64] genefilter_1.71.0          
##  [65] S4Vectors_0.27.12          
##  [66] desc_1.2.0                 
##  [67] BiocGenerics_0.35.4        
##  [68] BiocParallel_1.23.2        
##  [69] DOT_0.1                    
##  [70] GenomeInfoDb_1.25.10       
##  [71] shinydashboardPlus_0.7.5   
##  [72] rlang_0.4.7                
##  [73] pkgconfig_2.0.3            
##  [74] matrixStats_0.56.0         
##  [75] bitops_1.0-6               
##  [76] evaluate_0.14              
##  [77] lattice_0.20-41            
##  [78] purrr_0.3.4                
##  [79] htmlwidgets_1.5.1          
##  [80] bit_4.0.4                  
##  [81] tidyselect_1.1.0           
##  [82] bookdown_0.20              
##  [83] DESeq2_1.29.8              
##  [84] R6_2.4.1                   
##  [85] IRanges_2.23.10            
##  [86] generics_0.0.2             
##  [87] DelayedArray_0.15.7        
##  [88] DBI_1.1.0                  
##  [89] pillar_1.4.6               
##  [90] withr_2.2.0                
##  [91] survival_3.2-3             
##  [92] RCurl_1.98-1.2             
##  [93] tibble_3.0.3               
##  [94] crayon_1.3.4               
##  [95] shinyWidgets_0.5.3         
##  [96] shinyFiles_0.8.0           
##  [97] plotly_4.9.2.1             
##  [98] rhandsontable_0.3.7        
##  [99] rmarkdown_2.3              
## [100] locfit_1.5-9.4             
## [101] grid_4.1.0                 
## [102] data.table_1.13.0          
## [103] shinytoastr_2.1.1          
## [104] blob_1.2.1                 
## [105] digest_0.6.25              
## [106] xtable_1.8-4               
## [107] tidyr_1.1.1                
## [108] httpuv_1.5.4               
## [109] openssl_1.4.2              
## [110] stats4_4.1.0               
## [111] munsell_0.5.0              
## [112] viridisLite_0.3.0          
## [113] shinyjs_1.1                
## [114] askpass_1.1</code></pre>
</div>
<div id="funding" class="section level1">
<h1 class="hasAnchor">
<a href="#funding" class="anchor"></a>Funding</h1>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://lezhang.me/">Le Zhang</a>, <a href="https://www.dcassol.com/">Daniela Cassol</a>, <a href="https://github.com/mathrj">Ponmathi Ramasamy</a>, <a href="https://github.com/jianhaizhang">Jianhai Zhang</a>, <a href="https://github.com/gdmosher">Gordon Mosher</a>, <a href="http://girke.bioinformatics.ucr.edu/">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  
</body>
</html>
